lib/Auth/Source/Ldap.php

```diff
@@ -49,11 +67,182 @@ public function __construct(array $info, array $config)
-     * param array $sasl_arg  Associative array of SASL options
      * @return array  Associative array with the users attributes.
      */
-    protected function login(string $username, string $password, array $sasl_args = null): array
+    protected function login(string $username, string $password): array
+    {
+        $encryption = $this->ldapConfig->getString('encryption', 'ssl');
+        Assert::oneOf($encryption, ['none', 'ssl', 'tls']);
+
+        $version = $this->ldapConfig->getInteger('version', 3);
+        Assert::positiveInteger($version);
+
+        $ldapServers = [];
+        foreach (explode(' ', $this->ldapConfig->getString('connection_string')) as $connection_string) {
+            Assert::regex($connection_string, '#^ldap[s]?:\/\/#');
+
+            $ldap_servers[] = LdapObject::create(
+                $this->ldapConfig->getString('extension', 'ext_ldap'),
+                [
+                    'connection_string' => $connection_string,
+                    'encryption' => 'ssl',
+                    'version' => $version,
+                ]
+            );
+        }
+
+        $searchScope = $this->ldapConfig->getString('search.scope', Query::SCOPE_SUB);
+        Assert::oneOf($searchScope, [Query::SCOPE_BASE, Query::SCOPE_ONE, Query::SCOPE_SUB]);
+
+        $referrals = $this->ldapConfig->getValue('referrals', Query::DEREF_NEVER);
+        Assert::oneOf($referrals, [Query::DEREF_ALWAYS, Query::DEREF_NEVER, Query::DEREF_FINDING, Query::DEREF_SEARCHING]);
+
+        $timeout = $this->ldapConfig->getString('timeout', 3);
+        $searchBase = $this->ldapConfig->getArray('search.base');
+        $options = [
+            'scope' => $searchScope,
+            'timeout' => $timeout,
+            'deref' => $referrals,
+        ];
+
+        $searchEnable = $this->ldapConfig->getBoolean('search.enable', false);
+        if ($searchEnable === false) {
+            $dnPattern = $this->ldapConfig->getString('dnpattern');
+            $dn = str_replace('%username%', $username, $dnPattern);
+
+            $filter = '';
+        } else {
+            $searchUsername = $this->ldapConfig->getString('search.username');
+            Assert::notWhitespaceOnly($searchUsername);
+
+            $searchPassword = $this->ldapConfig->getString('search.password', null);
+            Assert::nullOrnotWhitespaceOnly($searchPassword);
+
+            $searchAttributes = $this->ldapConfig->getArray('search.attributes');
+            $searchFilter = $this->ldapConfig->getString('search.filter', null);
+
+            $ldap = $this->bind($ldapServers, $searchUsername, $searchPassword);
+
+            $filter = '';
+            foreach ($searchAttributes as $attr) {
+                $filter .= '(' . $attr . '=' . $username . ')';
+            }
+            $filter = '(|' . $filter . ')';
+
+            // Append LDAP filters if defined
+            if ($searchFilter !== null) {
+                $filter = "(&" . $filter . "" . $searchFilter . ")";
```

Comment:
What's with the empty string?

Reply:
I actually copied it like that, but it is unnecessary.. Removed it